### ISSUE TITLE
[BUG] Filtering a `ChoiceMap` with an internal `VectorChoiceMap` issue

### ISSUE THREAD

**Issue by ahiser1117** [genjax-community/genjax#954]
**Describe the bug**
When a hierarchical `GenerativeFunction` results in a `VectorChoiceMap` being inside another `ChoiceMap`, filtering the outer `ChoiceMap` to the point where the `VectorChoiceMap` is empty results in weird behavior.

**To Reproduce**

Here's the model definition. I think this is less important but I'll include it for completeness:

```python
import genjax
import jax
import jax.numpy as jnp
from scipy.stats import norm

from genjax import GenerativeFunction

def noiseless_model(a, b, y0, xs):
    def noiseless_model_update(activity_prev, x):
        act_t = (a * x + b * activity_prev)
        return act_t, act_t

    return jax.lax.scan(noiseless_model_update, y0, xs)[1]

def model_gen(xs) -> GenerativeFunction:
   
    @genjax.map_combinator(in_axes=(0, None, None, None, None))
    @genjax.static_gen_fn
    def kernel(xs, a, α, σ_SE, σ_noise):
        ζ0 = genjax.normal(0.0, 1.0) @ "ζ0"
        b = genjax.normal(0.0, 1.0) @ "b"
        y0 = genjax.normal(0.0, 1.0) @ "y0"

        ζ = jnp.exp(ζ0/5)

        z = noiseless_model(a, b, y0, xs)

        y = genjax.lgssm(z, α, σ_SE * ζ, σ_noise * ζ) @ "y"
        return y
    
    @genjax.static_gen_fn
    def model():
        a = genjax.normal(0.0, 1.0) @ "a" 
        
        α0 = genjax.normal(2.0, 1.0) @ "α0" 
        σ0_SE = genjax.normal(0.0, 1.0) @ "σ0_SE" 
        σ0_noise = genjax.normal(0.0, 1.0) @ "σ0_noise" 
        
        α = norm.cdf(α0)
        σ_SE = 0.5 * jnp.exp(σ0_SE)
        σ_noise = 0.125 * jnp.exp(σ0_noise * 0.5)

        ys = kernel(xs, a, α, σ_SE, σ_noise) @ "ys" 

        return ys
    
    return model

key = jax.random.PRNGKey(314159)

key, subkey = jax.random.split(key)
model = model_gen(jax.random.normal(subkey, (10, 50)))

key, subkey = jax.random.split(key)
tr = model.simulate(subkey, ())
```

If I then look at the resulting `ChoiceMap` from this `Trace` I have,
```python
chm = tr.strip()
console.print(chm)
```
```
(HierarchicalChoiceMap)
├── :α0
│   └── (ValueChoice)
│       └──  f32[]
├── :ys
│   └── (VectorChoiceMap)
│       └── (HierarchicalChoiceMap)
│           ├── :y0
│           │   └── (ValueChoice)
│           │       └──  f32[10]
│           ├── :b
│           │   └── (ValueChoice)
│           │       └──  f32[10]
│           ├── :ζ0
│           │   └── (ValueChoice)
│           │       └──  f32[10]
│           └── :y
│               └── (ValueChoice)
│                   └──  f32[10,50]
├── :σ0_SE
│   └── (ValueChoice)
│       └──  f32[]
├── :a
│   └── (ValueChoice)
│       └──  f32[]
└── :σ0_noise
    └── (ValueChoice)
        └──  f32[]
```

If I then filter this `ChoiceMap` to select for an argument that's not in the `VectorChoiceMap` the `VectorChoiceMap` is still there but has an `EmptyChoice` inner.

```python
selection = genjax.select("a")
filtered = chm.filter(selection)
console.print(filtered)
```
```
(HierarchicalChoiceMap)
├── :ys
│   └── (VectorChoiceMap)
│       └── (EmptyChoice)
└── :a
    └── (ValueChoice)
        └──  f32[]
```

If I then try to `console.print` this filtered `ChoiceMap` I get an error:
```python
console.print(filtered.get_selection())
```
```
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Cell In[11], line 1
----> 1 console.print(filtered.get_selection())

    [... skipping hidden 1 frame]

File ~/.conda/envs/genjax-py311/lib/python3.11/site-packages/genjax/_src/core/datatypes/generative.py:1342, in HierarchicalChoiceMap.get_selection(self)
   1340 trie = Trie()
   1341 for k, v in self.get_submaps_shallow():
-> 1342     trie = trie.trie_insert(k, v.get_selection())
   1343 return HierarchicalSelection(trie)

    [... skipping hidden 1 frame]

File ~/.conda/envs/genjax-py311/lib/python3.11/site-packages/genjax/_src/generative_functions/combinators/vector/vector_datatypes.py:232, in VectorChoiceMap.get_selection(self)
    230 subselection = self.inner.get_selection()
    231 # Static: get the leading dimension size value.
--> 232 dim = Pytree.static_check_tree_leaves_have_matching_leading_dim(
    233     self.inner,
    234 )
    235 return IndexedSelection(jnp.arange(dim), subselection)

File ~/.conda/envs/genjax-py311/lib/python3.11/site-packages/genjax/_src/core/pytree.py:162, in Pytree.static_check_tree_leaves_have_matching_leading_dim(tree)
    160 leaf_lengths = set(leaves)
    161 # all the leaves must have the same first dim size.
--> 162 assert len(leaf_lengths) == 1
    163 max_index = list(leaf_lengths).pop()
    164 return max_index

AssertionError: 
```

**Expected behavior**

I expect the resulting `ChoiceMap` stored in `filtered` to only contain the address `a` and the `VectorChoiceMap` to no longer be included.

**Comment by femtomc** [comment]
Possibly closed by #1018.

**Comment by sritchie** [comment]
When I try and reproduce this today on nightly I see

```
---------------------------------------------------------------------------
TracerArrayConversionError                Traceback (most recent call last)
Cell In[1], line 55
     52 model = model_gen(jax.random.normal(subkey, (10, 50)))
     54 key, subkey = jax.random.split(key)
---> 55 tr = model.simulate(subkey, ())

    [... skipping hidden 1 frame]

File <@beartype(genjax._src.generative_functions.static.static_gen_fn.StaticGenerativeFunction.simulate) at 0x15f186fc0>:55, in simulate(__beartype_object_5245105280, __beartype_get_violation, __beartype_conf, __beartype_object_5368311648, __beartype_func, *args, **kwargs)

File ~/code/python/genjax/src/genjax/_src/generative_functions/static/static_gen_fn.py:129, in StaticGenerativeFunction.simulate(self, key, args)
    120 @typecheck
    121 def simulate(
    122     self,
    123     key: PRNGKey,
    124     args: Tuple,
    125 ) -> StaticTrace:
    126     syntax_sugar_handled = push_trace_overload_stack(
    127         handler_trace_with_static, self.source
    128     )
--> 129     (args, retval, address_choices, score), cache_state = simulate_transform(
    130         syntax_sugar_handled
    131     )(key, args)
    132     return StaticTrace(
    133         self,
    134         args,
   (...)
    138         score,
    139     )

File ~/code/python/genjax/src/genjax/_src/generative_functions/static/static_transforms.py:286, in simulate_transform.<locals>.wrapper(key, args)
    283 @functools.wraps(source_fn)
    284 def wrapper(key, args):
    285     stateful_handler = SimulateHandler(key)
--> 286     retval = forward(source_fn)(stateful_handler, *args)
    287     (
    288         address_choices,
    289         cache_state,
    290         score,
    291     ) = stateful_handler.yield_state()
    292     return (
    293         args,
    294         retval,
    295         address_choices,
    296         score,
    297     ), cache_state

File <@beartype(genjax._src.core.interpreters.forward.forward.wrapped) at 0x2bcedaac0>:32, in wrapped(__beartype_object_5089698416, __beartype_get_violation, __beartype_conf, __beartype_func, *args, **kwargs)

File ~/code/python/genjax/src/genjax/_src/core/interpreters/forward.py:251, in forward.<locals>.wrapped(stateful_handler, *args)
    247 @functools.wraps(f)
    248 @typecheck
    249 def wrapped(stateful_handler: StatefulHandler, *args):
    250     interpreter = ForwardInterpreter()
--> 251     return interpreter.run_interpreter(
    252         stateful_handler,
    253         f,
    254         *args,
    255     )

    [... skipping hidden 1 frame]

File ~/code/python/genjax/src/genjax/_src/core/interpreters/forward.py:234, in ForwardInterpreter.run_interpreter(self, stateful_handler, fn, *args, **kwargs)
    231 def _inner(*args):
    232     return fn(*args, **kwargs)
--> 234 _closed_jaxpr, (flat_args, _, out_tree) = stage(_inner)(*args)
    235 _jaxpr, consts = _closed_jaxpr.jaxpr, _closed_jaxpr.literals
    236 flat_out = self._eval_jaxpr_forward(
    237     stateful_handler,
    238     _jaxpr,
    239     consts,
    240     flat_args,
    241 )

File ~/code/python/genjax/src/genjax/_src/core/interpreters/staging.py:81, in stage.<locals>.wrapped(*args, **kwargs)
     79 flat_fun, out_tree = api_util.flatten_fun_nokwargs(fun, in_tree)
     80 flat_avals = safe_map(get_shaped_aval, flat_args)
---> 81 typed_jaxpr = cached_stage_dynamic(flat_fun, tuple(flat_avals))
     82 return typed_jaxpr, (flat_args, in_tree, out_tree)

File ~/code/python/genjax/.venv/lib/python3.11/site-packages/jax/_src/linear_util.py:350, in cache.<locals>.memoized_fun(fun, *args)
    348   fun.populate_stores(stores)
    349 else:
--> 350   ans = call(fun, *args)
    351   if explain and config.explain_cache_misses.value:
    352     explain(fun.f, cache is new_cache, cache, key, ans)

File ~/code/python/genjax/src/genjax/_src/core/interpreters/staging.py:68, in cached_stage_dynamic(flat_fun, in_avals)
     66 @lu.cache
     67 def cached_stage_dynamic(flat_fun, in_avals):
---> 68     jaxpr, _, consts = pe.trace_to_jaxpr_dynamic(flat_fun, in_avals)
     69     typed_jaxpr = jc.ClosedJaxpr(jaxpr, consts)
     70     return typed_jaxpr

File ~/code/python/genjax/.venv/lib/python3.11/site-packages/jax/interpreters/partial_eval.py:103, in trace_to_jaxpr_dynamic(fun, in_avals, debug_info, keep_inputs)
    102 def trace_to_jaxpr_dynamic(fun, in_avals, debug_info=None, *, keep_inputs=None):  # noqa
--> 103   jaxpr, out_avals, consts, () = _trace_to_jaxpr_dynamic(
    104       fun, in_avals, debug_info, keep_inputs=keep_inputs)
    105   return jaxpr, out_avals, consts

File ~/code/python/genjax/.venv/lib/python3.11/site-packages/jax/_src/profiler.py:335, in annotate_function.<locals>.wrapper(*args, **kwargs)
    332 @wraps(func)
    333 def wrapper(*args, **kwargs):
    334   with TraceAnnotation(name, **decorator_kwargs):
--> 335     return func(*args, **kwargs)
    336   return wrapper

File ~/code/python/genjax/.venv/lib/python3.11/site-packages/jax/_src/interpreters/partial_eval.py:2322, in trace_to_jaxpr_dynamic(fun, in_avals, debug_info, keep_inputs)
   2320 with core.new_main(DynamicJaxprTrace, dynamic=True) as main:  # type: ignore
   2321   main.jaxpr_stack = ()  # type: ignore
-> 2322   jaxpr, out_avals, consts, attrs_tracked = trace_to_subjaxpr_dynamic(
   2323     fun, main, in_avals, keep_inputs=keep_inputs, debug_info=debug_info)
   2324   del main, fun
   2325 return jaxpr, out_avals, consts, attrs_tracked

File ~/code/python/genjax/.venv/lib/python3.11/site-packages/jax/_src/interpreters/partial_eval.py:2344, in trace_to_subjaxpr_dynamic(fun, main, in_avals, keep_inputs, debug_info)
   2342 in_tracers = _input_type_to_tracers(trace.new_arg, in_avals)
   2343 in_tracers_ = [t for t, keep in zip(in_tracers, keep_inputs) if keep]
-> 2344 ans = fun.call_wrapped(*in_tracers_)
   2345 out_tracers = map(trace.full_raise, ans)
   2346 jaxpr, consts, attrs_tracked = frame.to_jaxpr(out_tracers)

File ~/code/python/genjax/.venv/lib/python3.11/site-packages/jax/_src/linear_util.py:192, in WrappedFun.call_wrapped(self, *args, **kwargs)
    189 gen = gen_static_args = out_store = None
    191 try:
--> 192   ans = self.f(*args, **dict(self.params, **kwargs))
    193 except:
    194   # Some transformations yield from inside context managers, so we have to
    195   # interrupt them before reraising the exception. Otherwise they will only
    196   # get garbage-collected at some later time, running their cleanup tasks
    197   # only after this exception is handled, which can corrupt the global
    198   # state.
    199   while stack:

File ~/code/python/genjax/src/genjax/_src/core/interpreters/forward.py:232, in ForwardInterpreter.run_interpreter.<locals>._inner(*args)
    231 def _inner(*args):
--> 232     return fn(*args, **kwargs)

File ~/code/python/genjax/src/genjax/_src/generative_functions/supports_callees.py:46, in push_trace_overload_stack.<locals>.wrapped(*args)
     44 def wrapped(*args):
     45     GLOBAL_TRACE_HANDLER_STACK.append(handler)
---> 46     ret = fn(*args)
     47     GLOBAL_TRACE_HANDLER_STACK.pop()
     48     return ret

Cell In[1], line 39, in model_gen.<locals>.model()
     36 σ0_SE = genjax.normal(0.0, 1.0) @ "σ0_SE"
     37 σ0_noise = genjax.normal(0.0, 1.0) @ "σ0_noise"
---> 39 α = norm.cdf(α0)
     40 σ_SE = 0.5 * jnp.exp(σ0_SE)
     41 σ_noise = 0.125 * jnp.exp(σ0_noise * 0.5)

File ~/code/python/genjax/.venv/lib/python3.11/site-packages/scipy/stats/_distn_infrastructure.py:2067, in rv_continuous.cdf(self, x, *args, **kwds)
   2045 """
   2046 Cumulative distribution function of the given RV.
   2047
   (...)
   2064
   2065 """
   2066 args, loc, scale = self._parse_args(*args, **kwds)
-> 2067 x, loc, scale = map(asarray, (x, loc, scale))
   2068 args = tuple(map(asarray, args))
   2069 _a, _b = self._get_support(*args)

File ~/code/python/genjax/.venv/lib/python3.11/site-packages/jax/_src/core.py:684, in Tracer.__array__(self, *args, **kw)
    683 def __array__(self, *args, **kw):
--> 684   raise TracerArrayConversionError(self)

TracerArrayConversionError: The numpy.ndarray conversion method __array__() was called on traced array with shape float32[].
See https://jax.readthedocs.io/en/latest/errors.html#jax.errors.TracerArrayConversionError
```
